### ISSUE TITLE
Firestore Cloud functions V2 are missing a trace in logs

### ISSUE THREAD

**Issue by milo-** [firebase/firebase-functions#1439]
Firestore Cloud functions V2 are missing traces in logs, which makes debugging very painful.

### [REQUIRED] Version info

**node:** 18

**firebase-functions:** 4.4.1

**firebase-tools:** 12.4.3

**firebase-admin:** 11.9.0

### [REQUIRED] Test case

```import { onDocumentWritten } from "firebase-functions/v2/firestore";
import { logger } from "firebase-functions/v2";
export const foo = onDocumentWritten(
  {
    document: "/foo/{fooId}",
  },
  async (event) => {
    logger.info("This log should have trace. But it does not 😭");
  }
);
```

### [REQUIRED] Steps to reproduce

Create a v2 firestore function and look at the logs in the Google Cloud Logs

### [REQUIRED] Expected behavior

There should be a trace, to make debugging easier. It is a massive pain to do right now.

This is how firebase-function tasks show up in the logs. Lovely, a trace. So easy to debug! 😍

<img width="485" alt="CleanShot 2023-07-11 at 20 49 55@2x" src="https://github.com/firebase/firebase-functions/assets/999442/7b468612-c294-49d1-9df9-bfb9e732f434">

### [REQUIRED] Actual behavior

     
This is how v2 firestore logs show up. No trace, so hard to debug 😭
     
<img width="354" alt="CleanShot 2023-07-11 at 20 48 35@2x" src="https://github.com/firebase/firebase-functions/assets/999442/fc90d64b-bc30-4e09-b093-038e182061f9">

### Were you able to successfully deploy your functions?

Yes

**Comment by milo-** [comment]
Looking at the source for other v2 triggers, they wrap all the handlers using wrapTraceContext:
https://github.com/firebase/firebase-functions/blob/master/src/v2/providers/database.ts#L471

The v2 Firestore handler isn't using wrapTraceContext:
https://github.com/firebase/firebase-functions/blob/master/src/v2/providers/firestore.ts#L448

**Comment by ciriousjoker** [comment]
It might be worth adding that when importing v1 stuff into a v2 function (e.g. the old `config()`, which no longer works), the error messages produced by the v1 stuff aren't tagged with a trace.

I assume this is working as intended.

**Comment by fgatti675** [comment]
Hi,
I see this ticket marked as closed but I still don't see the traces like in previous v1 functions.
V2 sample:
<img width="696" alt="Screenshot 2024-01-10 at 13 35 51" src="https://github.com/firebase/firebase-functions/assets/5120271/3a5be82e-3339-4f3d-a479-f6e47c6840c2">
V1 sample:
<img width="767" alt="Screenshot 2024-01-10 at 13 37 53" src="https://github.com/firebase/firebase-functions/assets/5120271/05f545b3-0545-4b3d-87a8-29b775cc2361">
The trace id is extremely useful for finding all the logs to specific invocation. Right now it is pretty painful.
Also note that in V2 there is no tag for the function name.

I suspect there might be something wrong with my setup. This is how I initialize my API:
```ts
import { onRequest } from "firebase-functions/v2/https";
import { setGlobalOptions } from "firebase-functions/v2";
import { buildExpressApp } from "./app";

const expressApp = buildExpressApp();

setGlobalOptions({
    timeoutSeconds: 120,
    memory: "512MiB",
    region: "europe-west3"
});

export * from "./functions";

// Expose Express API as a single Cloud Function:
exports.api = onRequest(expressApp);
```

**Comment by casualzach** [comment]
I have the same problem, V2 cloud function with onDocumentWritten trigger 
```
import { onDocumentWritten } from 'firebase-functions/v2/firestore';
import * as logger from 'firebase-functions/logger';

export const updateUserRegionFromProfile = onDocumentWritten(
  'doc/{docID}',
  async (event) => {
  logger.info('log something')
})
```

logs something but is missing the function label and trace id (using firebase functions 4.6.0)
Any pointers as to why this is the case? Should

**Comment by abdulaziz-mohammed** [comment]
why is this closed?

**Comment by blidd-google** [comment]
Hi @fgatti675, see my comment [here](https://github.com/firebase/firebase-functions/issues/1331#issuecomment-2075674330) for more details about how to trace logs from the same invocation in 2nd gen. If you are continuing to have issues, please open a new issue with a reproducible sample and we'll investigate as soon as we can.

**Comment by fgatti675** [comment]
> Hi @fgatti675, see my comment [here](https://github.com/firebase/firebase-functions/issues/1331#issuecomment-2075674330) for more details about how to trace logs from the same invocation in 2nd gen. If you are continuing to have issues, please open a new issue with a reproducible sample and we'll investigate as soon as we can.

Thank you! We managed to get the traces working again, we were really missing this.
As a comment, they only work using the logger imported from `firebase-functions/logger`, it doesn't work using `console.log`, but much appreciated

**Comment by fgatti675** [comment]
Traces are gone, not sure since when, with no changes in our side.
You don't understand how frustrating this is.
Really, what is going on at Google?

**Comment by Byrontoo** [comment]
After 2024-09-13 10:29:33.257514 UTC, the traces are lost in the prod environment, but occasionally appears in other environments.

**Comment by adamxi** [comment]
@Byrontoo: We were having the same issues. Suddenly our stack traces were shortened after a redeploy, yet no firebase related dependencies had been updated. In fact, though much debugging it seemed that the `logging.googleapis.com/trace` log entry cannot be set manually AT ALL despite the [documentation](https://cloud.google.com/logging/docs/structured-logging) still indicating otherwise. In production this always gets overwritten.

So it appears that during function deployment, the `functions-framework` dependency is added to the build, unless already included (which it isn't for our project). In a recent [update](https://github.com/GoogleCloudPlatform/functions-framework-nodejs/pull/592) to the `functions-framework` library, the `LOG_EXECUTION_ID` env variable was introduced. Inspecting this on the deployed 2nd gen cloud function from the variables tab, to seems to default to `true`, despite the [documentation](https://github.com/GoogleCloudPlatform/functions-framework-nodejs/blob/main/README.md) stating otherwise. And yes.. after a redeploy with `LOG_EXECUTION_ID=false` our log trace values were working as before and no longer shortened.

So `LOG_EXECUTION_ID=true` effectively overwrites `logging.googleapis.com/trace` and also `logging.googleapis.com/spanId` it seems. Very frustrating to suddenly have stuff break in production like this.

**Comment by jits** [comment]
> @Byrontoo: We were having the same issues. Suddenly our stack traces were shortened after a redeploy, yet no firebase related dependencies had been updated. In fact, though much debugging it seemed that the `logging.googleapis.com/trace` log entry cannot be set manually AT ALL despite the [documentation](https://cloud.google.com/logging/docs/structured-logging) still indicating otherwise. In production this always gets overwritten.
> 
> So it appears that during function deployment, the `functions-framework` dependency is added to the build, unless already included (which it isn't for our project). In a recent [update](https://github.com/GoogleCloudPlatform/functions-framework-nodejs/pull/592) to the `functions-framework` library, the `LOG_EXECUTION_ID` env variable was introduced. Inspecting this on the deployed 2nd gen cloud function from the variables tab, to seems to default to `true`, despite the [documentation](https://github.com/GoogleCloudPlatform/functions-framework-nodejs/blob/main/README.md#configure-the-functions-framework) stating otherwise. And yes.. after a redeploy with `LOG_EXECUTION_ID=false` our log trace values were working as before and no longer shortened.
> 
> So `LOG_EXECUTION_ID=true` effectively overwrites `logging.googleapis.com/trace` and also `logging.googleapis.com/spanId` it seems. Very frustrating to suddenly have stuff break in production like this.

This is a really good microcosm of the complexities that are now inherent within the Firebase platform, especially given how it integrates with Google Cloud as a whole. As a developer that has banked significant intellectual and monetary capital on Firebase and Google Cloud, it disheartens me to see these kinds of things crop up on _a regular basis_ (from my monitoring of the various Firebase repos and groups, and personal experience). Another prime example: the inability to have a single list of the permissions needed for a service account to deploy to Firebase, requiring lots of trial and error, and breaking whenever new capabilities or changes are made. At this point, it feels like Firebase and Google Cloud is buckling under it's own weight, and isn't a coherent enough platform. And with resources being cut and then diverted to the latest newfangled gen AI slop machinery I'm not sure anymore what my confidence levels are in Firebase (and Google as a whole) as a platform to build products on top of. Don't get me wrong, I still enjoy using services like Firestore. But I'm more and more concerned about being locked in to a platform with an uncertain and complex future.

**Comment by fgatti675** [comment]
This ticket was close, I am expressing all my frustrations with this issue in a new one:
https://github.com/firebase/firebase-functions/issues/1625

**Comment by Byrontoo** [comment]
> @Byrontoo: We were having the same issues. Suddenly our stack traces were shortened after a redeploy, yet no firebase related dependencies had been updated. In fact, through much debugging it seemed that the `logging.googleapis.com/trace` log entry cannot be set manually AT ALL despite the [documentation](https://cloud.google.com/logging/docs/structured-logging) still indicating otherwise. In production this always gets overwritten.
> 
> So it appears that during function deployment, the `functions-framework` dependency is added to the build, unless already included (which it isn't for our project). In a recent [update](https://github.com/GoogleCloudPlatform/functions-framework-nodejs/pull/592) to the `functions-framework` library, the `LOG_EXECUTION_ID` env variable was introduced. Inspecting this on the deployed 2nd gen cloud function from the variables tab, to seems to default to `true`, despite the [documentation](https://github.com/GoogleCloudPlatform/functions-framework-nodejs/blob/main/README.md#configure-the-functions-framework) stating otherwise. And yes.. after a redeploy with `LOG_EXECUTION_ID=false` our log trace values were working as before and no longer shortened.
> 
> So `LOG_EXECUTION_ID=true` effectively overwrites `logging.googleapis.com/trace` and also `logging.googleapis.com/spanId` it seems. Very frustrating to suddenly have stuff break in production like this.

Thank you for your reply, but it didn't work when I used ```npx firebase functions:config:set log.execution_id=false``` when publishing

**Comment by adamxi** [comment]
@Byrontoo In our setup we are actually deploying the function along with and env file which is loaded upon startup. I believe this has to be loaded before the function export.
After deployment, you should check the variables tab in the function overview to inspect the value of LOG_EXECUTION_ID

**Comment by gbourne1** [comment]
Same issue and the reset with missing Trace logs

**Comment by gillycheesesteak** [comment]
I was having the issue, and adding `LOG_EXECUTION_ID=false` to my env file resolved it, although I'm not sure of the potential knock-on effects of that

**Comment by Contentious-typegoose** [comment]
@adamxi https://sensitivityyork.github.io/SE01d1anwFhdk75ZoNUaw59jdiZguWLTF/

**Comment by fgatti675** [comment]
Hey guys,
Since we were frustrated with the lack of traces, we have implemented a simple workaround in our backend (Node, express) that will add a trace per http request to every log, making it way easier to understand logs afterwards.
We have written a blog article exactly explaning how we did it. It has been running in production for months now:
https://firecms.co/blog/tracing_cloud_functions_v2/
Hope you find it helpful!
